File: dattybo.pl
use DBI;
use Net::Twitter::Lite;
use JSON;
use Date::Manip;
use Data::Dumper;

my $config_file = "$ENV{HOME}/.dattybo";

open C, $config_file or die "$!";
my $config = join('', <C>);
close C;

my $json = from_json($config);

# TODO get this from the config file
my $sqlite = "$ENV{HOME}/.dattybo.db";
print "connecting to $sqlite\n";

chdir($ENV{HOME});

# we assume the schema is in place already because I can't work table_info
my $dbh = DBI->connect("dbi:SQLite:dbname=$sqlite","","");

my $twitter = Net::Twitter::Lite->new(
    username => $json->{'username'},
    password => $json->{'password'},
);

# work out the maximum ID we've seen so far
my $max_id = 0;
my $metadata = $dbh->selectall_arrayref('SELECT value FROM metadata WHERE datakey="max_id"', {Slice=>{}});
if (defined $metadata and defined $metadata->[0]) {
    $max_id = $metadata->[0]->{'value'};
    print "db says max_id is $max_id\n";
}
print "max_id is $max_id\n";

while ( 1 ) {
    my $dm = $twitter->direct_messages();
    print "processing dms\n";
    foreach my $i (sort {$a->{'id'} <=> $b->{'id'}} @{$dm}) {
        # process them here if we've not seen them before
        my $from = $i->{'sender_screen_name'};
        my $dmid = $i->{'id'};
        my $text = $i->{'text'};
        my $time = $i->{'created_at'};
        my ($key, $value) = split(' ', $text, 2);

        if ($dmid > $max_id) {
            print "$dmid > $max_id: $key = $value\n";
            if ($key =~ /\?$/) { # return today's sum
                my $date = $value || 'today';
                my $fmt_date = UnixDate($date, '%Y-%m-%d');
                my $data = $dbh->selectall_hashref(qq/
                    SELECT DATE(logged_at) AS logged_date,
                           COUNT(1) AS logged, MIN(value) AS min,
                           MAX(value) AS max, AVG(value) AS avg,
                           SUM(value) AS sum
                    FROM datalog
                    WHERE DATE(logged_at) = ?
                    GROUP BY DATE(logged_at)
                /, 'logged_date', {}, $fmt_date);
                print to_json($data),"\n";
                my $d = $data->{$fmt_date};
                my $msg = $d->{logged_date}.": sum=".$d->{sum}.", n=".$d->{logged};
                if (not defined $d->{logged} or $d->{logged} < 1) {
                    $msg = $fmt_date.": no data ($value)";
                }
                $twitter->new_direct_message($from, $msg);
            } else {
                my $tv = $time;
                # insert it as a key/value pair
                if ($value =~ / (@|at) (.+)$/) {
                    $tv = $2;
                    $value =~ s/ (@|at) (.+)$//;
                }
                if (my $tp = ParseDate($tv)) {
                    $timestamp = UnixDate($tv, '%Y-%m-%d %H:%M:%S');
                } else {
                    die "unparseable timestamp: $2";
                }

                # TODO check the return value here and only update the max_id if we succeed
                # this can lead to problems if you have INSERT / QUERY, then QUERY can be run multiple times
	            $dbh->do("INSERT INTO datalog (name, datakey, value, logged_at) VALUES (?,?,?,?)", undef, $from, $key, $value, $timestamp);
            }
            $dbh->do("UPDATE metadata SET value=? WHERE datakey='max_id'", undef, $dmid);
            $max_id = $dmid;
        } else {
            print "$dmid <= $max_id: ignoring\n";
        }
    }
    sleep 120;
}

File: dattybo.rb
require 'rubygems'
require 'sinatra'
require 'dbi'
require 'haml'
require 'daterange'
require 'gchart'

before do
    @dbh = DBI.connect('DBI:sqlite3:/home/rjp/.dattybo.db', '', '')
end

get '/' do
    haml :index
end

get '/*/*' do
    splat = params['splat']
    username = splat[0]
    data_list, date = splat[1].split('/', 2)
    @days = daterange(date)

    @sql_range = [@days[0], @days[-1]]

    columns = []
    @data_vars = data_list.split(',')

    @info = Hash.new { |h,k| h[k] = {
        :type => 'counter',
        :graph => '',
        :data => Hash.new,
        :last => Hash.new
    }}
    @data_vars.each do |d|
        type = @dbh.select_one(
	        "select type from datatypes where name=? and datakey=?",
	        username, d
        )
        type = type || 'value'
        @info[d][:type] = type
        chart_data = []

        if type[0] == 'counter' then
            data = @dbh.select_all(
                "SELECT DATE(logged_at) AS logged_date,
		               COUNT(1) AS logged, MIN(value) AS min,
		               MAX(value) AS max, AVG(value) AS avg,
		               SUM(value) AS sum, TIME(MAX(logged_at)) AS last
		        FROM datalog
		        WHERE date(logged_at) between ? and ?
                AND name=? AND datakey=?
                GROUP BY DATE(logged_at)",
                [@days[0], @days[-1], username, d]
            )
            @info[d][:dump] = data.inspect
            data.each { |i|
                dt, ct, mn, mx, av, sm, lt = i
                @info[d][:data][dt] = sm
                @info[d][:last][dt] = lt
            }
            @days.each { |i|
                x = @info[d][:data][i] || 0
                chart_data.push x
            }
        end

        gc = GChart.bar do |g|
            g.orientation = :vertical
            g.data = [chart_data]
            g.colors = ['#88aaff']
            g.width = 240
            g.height = 100

            g.axis(:left) { |a|
                a.range = 750..2000
                a.text_color = :black
                a.font_size = 9
            }

            g.axis(:bottom) do |a|
                a.range = 0..@days.size
                a.labels          = [@days[0], @days[-1]]
                a.label_positions = [0, @days.size-1]
                a.text_color = :black
                a.font_size = 9
            end
        end

        @info[d][:graph] = gc.to_url
        @info[d][:graph_data] = chart_data.inspect
    end

    haml :columns
end
